File: csharp/Patterns/AbstractFactory.cs
using System;

namespace DesignPatterns.Patterns;

public static class AbstractFactory
{
    private interface IChair
    {
        string Sit();
    }

    private interface ISofa
    {
        string Nap();
    }

    private class ModernChair : IChair
    {
        public string Sit() => "Sit on a sleek modern chair";
    }

    private class ModernSofa : ISofa
    {
        public string Nap() => "Nap on a clean modern sofa";
    }

    private class VictorianChair : IChair
    {
        public string Sit() => "Sit on a fancy Victorian chair";
    }

    private class VictorianSofa : ISofa
    {
        public string Nap() => "Nap on a cozy Victorian sofa";
    }

    private interface IFurnitureFactory
    {
        IChair CreateChair();
        ISofa CreateSofa();
    }

    private class ModernFactory : IFurnitureFactory
    {
        public IChair CreateChair() => new ModernChair();
        public ISofa CreateSofa() => new ModernSofa();
    }

    private class VictorianFactory : IFurnitureFactory
    {
        public IChair CreateChair() => new VictorianChair();
        public ISofa CreateSofa() => new VictorianSofa();
    }

    private class LivingRoom
    {
        private readonly IChair _chair;
        private readonly ISofa _sofa;

        public LivingRoom(IFurnitureFactory factory)
        {
            _chair = factory.CreateChair();
            _sofa = factory.CreateSofa();
        }

        public string Relax() => $"{_chair.Sit()} and {_sofa.Nap()}";
    }

    public static void Run()
    {
        Console.WriteLine("\nAbstract Factory");
        foreach (var factory in new IFurnitureFactory[] { new ModernFactory(), new VictorianFactory() })
        {
            var room = new LivingRoom(factory);
            Console.WriteLine(room.Relax());
        }
    }
}

File: python/patterns/abstract_factory.py
from abc import ABC, abstractmethod


class Chair(ABC):
    @abstractmethod
    def sit(self) -> str:
        raise NotImplementedError


class Sofa(ABC):
    @abstractmethod
    def nap(self) -> str:
        raise NotImplementedError


class ModernChair(Chair):
    def sit(self) -> str:
        return "Sit on a sleek modern chair"


class ModernSofa(Sofa):
    def nap(self) -> str:
        return "Nap on a clean modern sofa"


class VictorianChair(Chair):
    def sit(self) -> str:
        return "Sit on a fancy Victorian chair"


class VictorianSofa(Sofa):
    def nap(self) -> str:
        return "Nap on a cozy Victorian sofa"


class FurnitureFactory(ABC):
    @abstractmethod
    def create_chair(self) -> Chair:
        raise NotImplementedError

    @abstractmethod
    def create_sofa(self) -> Sofa:
        raise NotImplementedError


class ModernFactory(FurnitureFactory):
    def create_chair(self) -> Chair:
        return ModernChair()

    def create_sofa(self) -> Sofa:
        return ModernSofa()


class VictorianFactory(FurnitureFactory):
    def create_chair(self) -> Chair:
        return VictorianChair()

    def create_sofa(self) -> Sofa:
        return VictorianSofa()


class LivingRoom:
    def __init__(self, factory: FurnitureFactory) -> None:
        self.chair = factory.create_chair()
        self.sofa = factory.create_sofa()

    def relax(self) -> str:
        return f"{self.chair.sit()} and {self.sofa.nap()}"


def run() -> None:
    print("\nAbstract Factory")
    for factory in (ModernFactory(), VictorianFactory()):
        room = LivingRoom(factory)
        print(room.relax())
